domain/domain.py

```diff
@@ -201,25 +207,119 @@ def generate_domain(self):
+        tic = time.clock()
+        pruned_domain = {}
+        for row in domain_df[['_tid_', 'attribute', 'domain']].to_records():
+            pruned_domain[row['_tid_']] = pruned_domain.get(row['_tid_'], {})
+            pruned_domain[row['_tid_']][row['attribute']] = row['domain'].split('|||')
+        estimator = RecurrentLogistic(self.ds, pruned_domain, self.active_attributes)
+        estimator.train(num_recur=1, num_epochs=3, batch_size=self.env['batch_size'])
+        logging.debug('DONE training posterior model in %.2fs', time.clock() - tic)
+
+        # Predict probabilities for all pruned domain values.
+        logging.debug('predicting domain value probabilities from posterior model...')
+        tic = time.clock()
+        # raw records indexed by tid
+        raw_records_by_tid = {row['_tid_']: row for row in records}
+        domain_records = domain_df.to_records()
+        preds_by_cell = estimator.predict_pp_batch(raw_records_by_tid, domain_records)
+        logging.debug('DONE predictions in %.2f secs, re-constructing cell domain...', time.clock() - tic)
+
+        logging.debug('re-assembling final cell domain table...')
+        tic = time.clock()
+        # iterate through raw/current data and generate posterior probabilities for
+        # weak labelling
+        num_weak_labels = 0
+        updated_domain_df = []
+        for preds, row in tqdm(zip(preds_by_cell, domain_records)):
+            # no need to modify single value cells
+            if row['fixed'] == CellStatus.SINGLE_VALUE.value:
+                updated_domain_df.append(row)
+                continue
+
+            # prune domain if any of the values are above our domain_prune_thresh
+            preds = [[val, proba] for val, proba in preds if proba >= self.domain_prune_thresh] or preds
+
+            # cap the maximum # of domain values to self.max_domain
+            domain_values = [val for val, proba in sorted(preds, key=lambda pred: pred[1], reverse=True)[:self.max_domain]]
+
+            # ensure the initial value is included
+            if row['init_value'] not in domain_values:
+                domain_values.append(row['init_value'])
+            # update our memoized domain values for this row again
+            row['domain'] = '|||'.join(domain_values)
+            row['domain_size'] = len(domain_values)
+            row['weak_label_idx'] = domain_values.index(row['weak_label'])
+            row['init_index'] = domain_values.index(row['init_value'])
+
+            # Assign weak label if domain value exceeds our weak label threshold
+            weak_label, weak_label_prob = max(preds, key=lambda pred: pred[1])
+
+            if weak_label_prob >= self.weak_label_thresh:
+                num_weak_labels+=1
+
+                weak_label_idx = domain_values.index(weak_label)
+                row['weak_label'] = weak_label
+                row['weak_label_idx'] = weak_label_idx
+                row['fixed'] = CellStatus.WEAK_LABEL.value
```

Comment:
At this point in the domain df, there may be rows where weak_label is set to some non-null value, but these cells might not be used as labelled data in the repair model because the weak_label_prob doesn't exceed weak label threshold. 

AFAICT this shouldn't cause any problems in the repair model because a cell is considered labelled training data if `fixed` is set to `CellStatus.WEAK_LABEL.value`. However, it might make more sense to set the weak_label for these values to null or some invalid value to avoid confusion.